src/Rooms.ts

```diff
@@ -57,23 +57,38 @@ export function guessAndSetDMRoom(room: Room, isDirect: boolean): Promise<void>
 /**
  * Marks or unmarks the given room as being as a DM room.
  * @param {string} roomId The ID of the room to modify
- * @param {string} userId The user ID of the desired DM
+ * @param {string | null} userId The user ID of the desired DM
  room target user or null to un-mark
  this room as a DM room
  * @returns {object} A promise
  */
-export async function setDMRoom(roomId: string, userId: string): Promise<void> {
+export async function setDMRoom(roomId: string, userId: string | null): Promise<void> {
     if (MatrixClientPeg.get().isGuest()) return;
 
     const mDirectEvent = MatrixClientPeg.get().getAccountData(EventType.Direct);
-    let dmRoomMap = {};
+    const currentContent = mDirectEvent?.getContent() || {};
 
-    if (mDirectEvent !== undefined) dmRoomMap = { ...mDirectEvent.getContent() }; // copy as we will mutate
+    // already marked as DM
+    if (userId && currentContent[userId]?.includes?.(roomId)) return;
+
+    if (userId === null) {
+        const roomInDMs = Object.values(currentContent).find((roomIds: any) => {
+            if (!Array.isArray(roomIds)) return false;
+            return roomIds.includes(roomId);
+        });
+
+        // skip remove unknown room
+        if (!roomInDMs) return;
+    }
+
+    let dmRoomMap: Map<string, string[]> = new Map();
+
+    if (mDirectEvent !== undefined) dmRoomMap = new Map(Object.entries(currentContent)); // copy as we will mutate
```

Comment:
given currentContent is valid either way, it feels like we may as well make this unconditional

```suggestion
    const dmRoomMap: Map<string, string[]> = new Map(Object.entries(currentContent)); // copy as we will mutate
```